### PR TITLE
[WIP] feat: Read all graphql configuration from postgres

### PR DESCRIPTION
**Pull request by david-martin**
feat: Read Data Source, Schema & Resolver configuration from posgres

The `sequelize` module is used for this.
A `DataSource`, `GraphQLSchema` & `Resolver` model is defined.
On startup, the first schema is retrieved from the database,
and all Resolvers & DataSources are read and passed into the schema parser
logic.

The local dev setup has changed due to the introduction of postgres.
See README for details.
The gist of it is:

```
npm run db:start
npm run db:seed      # Creates sequelize model tables & seeds some example data
npm run dev
```

Access graphiql on http://localhost:8000/graphiql

to stop:

```
npm run db:stop
```

*Other changes in this PR*

* remove example json files as they are no longer used
* remove k8s setup from README (to be replaced by a link to APB repo/setup in future)
* a change to the resolverMaker & datasourceParser logic to look for a `field` & `name` key respectively on each item in the item array, rather than using the key of an object

TODO:

* [x] Add models for GraphQLSchema, Resolver & Data Sources, removing all file based config
* [x] Merge with in-memory poc changes
* [x] Verify schema is OK for the Admin UI
* [ ] Tests

Separate follow up PR for:

* Auto reload the schema if the database table changes (any creates/deletes/updates)

### CODE REVIEW

File: models/dataSource.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports = (sequelize, DataTypes) => {
+  const DataSource = sequelize.define('DataSource', {
+    name: DataTypes.STRING,
+    type: DataTypes.ENUM('InMemory', 'Postgres'),
+    config: DataTypes.JSON
+  })
+
+  return DataSource
+}
```
Comment (david-martin) [inline models/dataSource.js]: @pb82 fyi, here is the sequelize model for a Data Source, based on the schema being used for the UI
(i.e. https://github.com/aerogear/data-sync-ui/blob/master/gql/schema.js#L4-L16)

Comment (pb82) [reply models/dataSource.js]: @david-martin thanks, we will use the same.

File: README.md [hunk 2]
```diff
@@ -10,7 +10,7 @@ GraphQL based data sync server for mobile, with backend integration capabilities
 
 ```
 docker run --rm --name=postgres -p 5432:5432 -e POSTGRES_PASSWORD=mysecretpassword -v `pwd`/examples:/tmp/examples -d postgres
-docker exec postgres psql -U postgres -f /tmp/examples/create_tables.example.sql
+docker exec postgres psql -U postgres -f /tmp/examples/create_example_datasource.sql
```
Comment (aliok) [inline README.md]: This fails because the table is not created before insert operation. @david-martin

Comment (david-martin) [reply README.md]: Good catch.
I'll update to start the server before creating the example datasource

Comment (aliok) [reply README.md]: So, what I understand is, Sequlize should create the table. But it didn't do in my case. After I manually created the table and run the server, I saw log statements like "CREATE TABLE etc " in the logs.

Here are the log statements:
```
Executing (default): SELECT "id", "name", "type", "config", "createdAt", "updatedAt" FROM "DataSources" AS "DataSource";
Executing (default): SELECT t.typname enum_name, array_agg(e.enumlabel ORDER BY enumsortorder) enum_value FROM pg_type t JOIN pg_enum e ON t.oid = e.enumtypid JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace WHERE n.nspname = 'public' AND t.typname='enum_DataSources_type' GROUP BY 1
Executing (default): CREATE TYPE "public"."enum_DataSources_type" AS ENUM('InMemory', 'Postgres');
Executing (default): SELECT typname, typtype, oid, typarray FROM pg_type WHERE (typtype = 'b' AND typname IN ('hstore', 'geometry', 'geography')) OR (typtype = 'e')
Executing (default): CREATE TABLE IF NOT EXISTS "DataSources" ("id"   SERIAL , "name" VARCHAR(255), "type" "public"."enum_DataSources_type", "config" JSON, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
Executing (default): SELECT i.relname AS name, ix.indisprimary AS primary, ix.indisunique AS unique, ix.indkey AS indkey, array_agg(a.attnum) as column_indexes, array_agg(a.attname) AS column_names, pg_get_indexdef(ix.indexrelid) AS definition FROM pg_class t, pg_class i, pg_index ix, pg_attribute a WHERE t.oid = ix.indrelid AND i.oid = ix.indexrelid AND a.attrelid = t.oid AND t.relkind = 'r' and t.relname = 'DataSources' GROUP BY i.relname, ix.indexrelid, ix.indisprimary, ix.indisunique, ix.indkey ORDER BY i.relname;
```
It first does a "SELECT" and in later stages it does "CREATE TABLE IF NOT EXISTS".

Comment (aliok) [reply README.md]: Other than this, progress so far looks good.

Comment (david-martin) [reply README.md]: @aliok I've updated the README steps, and changed the order on startup to fully initialise models first (i.e. create if not exists) before trying to list any.

i.e. this https://github.com/aerogear/data-sync-server/pull/9/commits/8f83102c13e4bafd78687e39b1a33f16e760c2ed

**Comment by darahayes** [comment]
@david-martin is this ready to go?

**Comment by david-martin** [comment]
@darahayes not quite.
About to push resolver & schema models too.
Also, I'd like to try get some tests in place.

File: models/resolver.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+module.exports = (sequelize, DataTypes) => {
```
Comment (david-martin) [inline models/resolver.js]: fyi @pb82 model for Resolver. See https://github.com/aerogear/data-sync-server/pull/9/files#diff-5709bb01e7a28a79644cde19374931d6R42 for example of what gets inserted

File: models/graphQLSchema.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module.exports = (sequelize, DataTypes) => {
```
Comment (david-martin) [inline models/graphQLSchema.js]: fyi @pb82 model for GraphQLSchema. See https://github.com/aerogear/data-sync-server/pull/9/files#diff-5709bb01e7a28a79644cde19374931d6R3 for example of what gets inserted

**Comment by david-martin** [comment]
@darahayes @aliok I've rebased and pushed a significant update to this.
Please see the PR description for info.

**Comment by david-martin** [comment]
@darahayes I have rebased on top of your in-memory datasource branch.
It's now possible to run the server (after starting & seeding the db),
and make requests using graphiql against the in memory data source.

**Comment by darahayes** [comment]
Closing this because changes were merged with #12

**Pull request closed without merging**